backend/src/v2/driver/driver.go

```diff
@@ -671,13 +671,22 @@ func resolveInputs(ctx context.Context, dag *metadata.DAG, iterationIndex *int,
 				}
 			}
 			value, hasValue := inputs.GetParameterValues()[name]
-			if !hasValue {
-				if spec.GetDefaultValue() == nil {
-					return fmt.Errorf("input parameter %q is required", name)
+
+			// Handle when parameter does not have input value
+			if !hasValue && inputsSpec.GetParameters()[name].IsOptional == false {
+				// when parameter is not optional and there is no input value, report error
+				return fmt.Errorf("no value provided for non-optional parameter %q", name)
+			} else if !hasValue && inputsSpec.GetParameters()[name].IsOptional == true {
+				// when parameter is optional and there is no input value, value comes from default value
+				if spec.GetDefaultValue() != nil {
+					inputs.GetParameterValues()[name] = spec.GetDefaultValue()
+					value = spec.GetDefaultValue()
+				} else {
```

Comment:
I don't recall if we reached consensus on this, but I called out in our group chat that this entire block can be omitted--only add value to executor input when users provided a value, leave default value to be filled in inside the container regardless whether it's `None` or something else. 
If you decide to keep this block at this moment. Can you please add a TODO comment above say we might want to revisit whether to have this block at all, so that we don't forgot all the context.

Reply:
To be clear, for Python-function component, whether we pass the default value or not makes no difference, and I was proposing to not pass. For customer component though, it depends on whether user code inside the container handles the default value. My argument was they should, thus passing default value here is not needed and should be avoided unless necessary. The other argument was some existing custom container may not work in this way, if we don't pass default value, it could break them. So I could see going either way at this moment. Starting with not passing default value is fine, and we could change it if too many users are on the other side.

Reply:
I agree, both sides have their reasonings. I will add a note here about both possible choices.